### DISCUSSION TITLE
Running Solution

### DISCUSSION THREAD

**Discussion by leninc** [Daniel-Krzyczkowski/MicrosoftAzure#16] [Q&A]
Hi I'm trying to run the solution in the asp-net-core-microservices-with-azure-and-docker branch, but i'm not able to, i can't find where i can edit the values for service bus or sql server or instructions on the file that's needed and where it would be read from. is there any file that explain the steps needed to be able to compile and run this locally with my azure settings?

**Comment by Daniel-Krzyczkowski** [comment]
Hi,
Unfortunately I did not prepare any file with exact configuration in the Azure cloud but:
1. [Here ](https://github.com/Daniel-Krzyczkowski/MicrosoftAzure/issues/12) you can find my explanation how to setup Service Bus
2. SQL connection string can be found below (once you create Azure SQL database in the Azure portal):
![sql01](https://user-images.githubusercontent.com/15244422/104548967-34419980-5632-11eb-9790-ffbf1e298326.PNG)

I hope above information will help you. Oh, one more thing. If you want to run solution on your local machine, please update the settings in the _appsettings.Development.json_ file in each API project (like [this ](https://github.com/Daniel-Krzyczkowski/MicrosoftAzure/blob/master/asp-net-core-microservices-with-azure-and-docker/src/CarsIsland/Catalog/CarsIsland.Catalog.API/appsettings.Development.json) one).

**Reply by leninc** [reply]
thanks for your detailed and quick reply, can i ask why you use the integration event and not any of the message class available either from azure.servicebus or windows.servicebus like, message, servicebusmessage or brokeredmessage?

**Reply by leninc** [reply]
Another question is  do these events

      services.AddSingleton<IEventBusSubscriptionsManager, InMemoryEventBusSubscriptionsManager>();
      services.AddTransient<IIntegrationEventHandler<CarPricePerDayChangedIntegrationEvent>, CarPricePerDayChangedIntegrationEventHandler>

take care of the handler being notified every time a new event has been published or is it needed to pool the topic for or subscription for that?

**Reply by Daniel-Krzyczkowski** [reply]
> thanks for your detailed and quick reply, can i ask why you use the integration event and not any of the message class available either from azure.servicebus or windows.servicebus like, message, servicebusmessage or brokeredmessage?

Integration event is an abstraction used to separate different events across different microservices. When new integration event is sent, underneath there is Azure Service Bus message sent. Please see [AzureServiceBusEventBus ](https://github.com/Daniel-Krzyczkowski/MicrosoftAzure/blob/master/asp-net-core-microservices-with-azure-and-docker/src/CarsIsland/BuildingBlocks/CarsIsland.EventBus/AzureServiceBusEventBus.cs) class imlementation. When new event is sent, there is new [message created](https://github.com/Daniel-Krzyczkowski/MicrosoftAzure/blob/b7c82eb7cae612e550637601cc2694c103afd9f6/asp-net-core-microservices-with-azure-and-docker/src/CarsIsland/BuildingBlocks/CarsIsland.EventBus/AzureServiceBusEventBus.cs#L50). Integration event classes were used to have consistent library of events that are understandable across different microservices.



> Another question is do these events
> 
> ```
>   services.AddSingleton<IEventBusSubscriptionsManager, InMemoryEventBusSubscriptionsManager>();
>   services.AddTransient<IIntegrationEventHandler<CarPricePerDayChangedIntegrationEvent>, CarPricePerDayChangedIntegrationEventHandler>
> ```
> 
> take care of the handler being notified every time a new event has been published or is it needed to pool the topic for or subscription for that?

[InMemoryEventBusSubscriptionsManager](https://github.com/Daniel-Krzyczkowski/MicrosoftAzure/blob/master/asp-net-core-microservices-with-azure-and-docker/src/CarsIsland/BuildingBlocks/CarsIsland.EventBus/InMemoryEventBusSubscriptionsManager.cs) is responsible for handling knowledge about subscriptions. Azure Service Bus do not provide any class that will provide you with the knowledge about event subscriptions. InMemoryEventBusSubscriptionsManager instance acts as a registry to keep information about subscriptions for different kind of events.
Then, there is an interface called [IIntegrationEventHandler](https://github.com/Daniel-Krzyczkowski/MicrosoftAzure/blob/master/asp-net-core-microservices-with-azure-and-docker/src/CarsIsland/BuildingBlocks/CarsIsland.EventBus/Events/Interfaces/IIntegrationEventHandler.cs) which is implemented in each microservice to handle specific events once they happen. Example of such event handler is [CarPricePerDayChangedIntegrationEventHandler ](https://github.com/Daniel-Krzyczkowski/MicrosoftAzure/blob/master/asp-net-core-microservices-with-azure-and-docker/src/CarsIsland/Reservation/CarsIsland.Reservation.API/Core/IntegrationEvents/EventHandlers/CarPricePerDayChangedIntegrationEventHandler.cs). You can see that this handler handles event related to price change for a specific car.

I hope above answer clarified a bit.

**Reply by leninc** [reply]
yes, between this and you youtube video i think i got what i need, thanks

**Reply by leninc** [reply]
InMemoryEventBusSubscriptionsManager how this class can be implemented if the services are in different repos and what happens in case of an outage or failure?

**Reply by Daniel-Krzyczkowski** [reply]
Very good question. Let's start with separate repositories for microservices. In this case you have to think about strategy for sharing common source code across all microservices. There are a few approaches. In my case I use [Azure Artifacts](https://docs.microsoft.com/en-us/azure/devops/artifacts/overview?view=azure-devops). With Azure Artifacts you can create and share Maven, npm, and NuGet package feeds from public and private sources. I share private NuGet packages with common source code.

When it comes to InMemoryEventBusSubscriptionsManager - believe me or not but recently I thought about it. If there is an outage - in the current setup all subscriptions will be lost. If you want to handle situations like this, you should provide some kind of persistence. You could store information about subscriptions for each service in the service like Azure Table Storage (to reduce the cost). Then, when there is a failure and microservice is restarted, configuration should be pulled from the external service.

**Comment by leninc** [comment]
Hi Daniel, how do you set up the users in your solutions, i'm testing my connection and everything is working except for users in redis that does not fetch any, so it does not process the update on the reservation side.

**Reply by Daniel-Krzyczkowski** [reply]
I would like to help but please provide more details because I am not sure if I understood you correctly. Do you mean you have problem with storing information about users in Redis cache related to reservations?

**Reply by leninc** [reply]
Sure, let me try to explain myself better, i see that the solutions check for users in redis but i don't have any and i haven't seen any file for seeding, how do the users get configure or stored in redis?

**Reply by Daniel-Krzyczkowski** [reply]
Oh got it thanks. No, at this point there is no file for seeding so unfortunately you will have to do it yourself. I did not fully complete implementation of this solution yet.
